maintenance/program/src/processor.rs

```diff
@@ -0,0 +1,317 @@
+            payer_info.clone()
+        ],
+        &[&signers_seeds[..]],
+    )?;
+
+    Ok(())
+}
+
+/// Processes Set Authority instruction
+pub fn process_set_authority(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let bpf_loader_program_info = next_account_info(account_info_iter)?; // 0
+    let maintained_program_info = next_account_info(account_info_iter)?; // 1
+    let maintained_program_data_info = next_account_info(account_info_iter)?; // 2
+    let maintenance_record_info = next_account_info(account_info_iter)?; // 3
+    let authority_info = next_account_info(account_info_iter)?; // 4
+    let new_authority_info = next_account_info(account_info_iter)?; // 5
+    let payer_info = next_account_info(account_info_iter)?; // 6
+
+    if !authority_info.is_signer {
+        return Err(MaintenanceError::MissingRequiredSigner.into());
+    }
+
+    let maintenance_seeds = get_maintenance_record_seeds(maintained_program_info.key);
+    let (_maintenance_address, bump_seed) = Pubkey::find_program_address(&maintenance_seeds, program_id);
+
+    let mut signers_seeds = maintenance_seeds.to_vec();
+    let bump = &[bump_seed];
+    signers_seeds.push(bump);
+
+    let upgrade_authority_instruction = set_upgrade_authority(
+        maintained_program_info.key,
+        maintenance_record_info.key,
+        Some(new_authority_info.key),
+    );
+
+    invoke_signed(
+        &upgrade_authority_instruction,
+        &[
+            bpf_loader_program_info.clone(),
+            maintained_program_info.clone(),
+            maintained_program_data_info.clone(),
+            maintenance_record_info.clone(),
+            new_authority_info.clone(),
+            payer_info.clone()
+        ],
+        &[&signers_seeds[..]],
+    )?;
+
+    Ok(())
+}
+
+/// Processes CloseMaintenance instruction
+pub fn process_close_maintenance(
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let maintenance_record_info = next_account_info(account_info_iter)?; // 0
+    let destination_account_info = next_account_info(account_info_iter)?; // 1
+    let program_data_info = next_account_info(account_info_iter)?; // 5
+
+    if maintenance_record_info.key == destination_account_info.key {
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    let program_authority: Pubkey = {
+        let program_data_offset: usize = UpgradeableLoaderState::programdata_data_offset().map_err(|_| MaintenanceError::BufferDataOffsetError )?;
+        let program_buffer: &[u8] = &program_data_info.data.borrow();
+        let upgrade_authority_data: &[u8] = program_buffer.get((program_data_offset-PUBKEY_BYTES)..program_data_offset).ok_or(MaintenanceError::BufferDataOffsetError)?;
+        Pubkey::try_from_slice(upgrade_authority_data).map_err(|_| MaintenanceError::AuthorityDeserializationError )?
```

Comment:
Try to deserialize data stored in the `programdata` account.
It seems that Solana BPFUgreadable uses this code to deserialize data: https://github.com/solana-labs/solana/blob/4aeb566c856cf904456cf31b20eef07c40e4be15/sdk/src/account.rs#L472-L476

Reply:
Fixed. bincode::deserialize(...) used.